src/CSVMsg.chpl

```diff
@@ -487,38 +492,356 @@ module CSVMsg {
+        if line.find('"') != -1 {
+            hasQuotes = true;
+            var totalQuotes = line.count('"');
+
+            // If odd number of quotes, keep reading until balanced
+            while totalQuotes % 2 == 1 {
+                try {
+                    line = reader.readLine(stripNewline=true);
+                    totalQuotes += line.count('"');
+                } catch e: EofError {
+                    break;
+                } catch e: UnexpectedEofError {
+                    break;
+                }
+            }
+        }
+
+        return hasQuotes;
+    }
+
+    // Full helper: read and store lines for complete CSV record (for readCSVRecord)
+    proc readCSVRecordLines(reader: fileReader(?)) : list(string) throws {
+        var lines: list(string);
+        var line: string;
+
+        // Read the first line
+        line = reader.readLine(stripNewline=true);
+        lines.pushBack(line);
+
+        // Quick check for quotes
+        if line.find('"') != -1 {
+            var totalQuotes = line.count('"');
+
+            // If odd number of quotes, keep reading until balanced
+            while totalQuotes % 2 == 1 {
+                try {
+                    line = reader.readLine(stripNewline=true);
+                    lines.pushBack(line);
+                    totalQuotes += line.count('"');
+                } catch e: EofError {
+                    break;
+                } catch e: UnexpectedEofError {
+                    break;
+                }
+            }
+        }
+
+        return lines;
+    }
+
+    // Skip a complete CSV record (handles multi-line quoted fields)
+    proc skipCSVRecord(reader: fileReader(?)) throws {
+        advanceCSVRecord(reader);
+        // Record is now completely skipped
+    }
+
+    // Fast row counting with quote detection - much faster than readCSVRecord
+    proc countRowsAndQuotes(reader: fileReader(?)) : (int, bool) throws {
+        var rowCount = 0;
+        var hasAnyQuotes = false;
+
+        try {
+            while true {
+                var hasQuotes = advanceCSVRecord(reader);
+                if hasQuotes {
+                    hasAnyQuotes = true;
+                }
+                rowCount += 1;
+            }
+        } catch e: EofError {
+            // Normal end of file
+        }
+
+        return (rowCount, hasAnyQuotes);
+    }
+
+    // Read a complete CSV record from the reader, handling multi-line quoted fields
+    proc readCSVRecord(reader: fileReader(?)) throws {
+        var lines: list(string);
+        try {
+            lines = readCSVRecordLines(reader);
+        } catch e: EofError {
+            throw e; // Re-throw EOF to signal end of file
+        } catch e: UnexpectedEofError {
+            throw e;
+        }
+
+        // Return single line if no multi-line record
+        if lines.size == 1 {
+            return lines[0];
+        } else {
+            return "\n".join(lines.toArray());
+        }
+    }
```

Comment:
as a nit, this code feels needlessly complex

```chapel
var lines = readCSVRecordLines(reader);
if lines.size == 1 { 
  return lines[0];
} else {
  return "\n".join(lines.toArray());
}
```

This should have the same behavior, no need to manually rethrow the exception. If you want to be more explicit you could do `var lines = try readCSVRecordLines(reader); `